tests/test_enum.rs

```diff
@@ -72,10 +72,59 @@ fn test_enum_eq_incomparable() {
     })
 }
 
+#[test]
+fn test_enum_compare_by_identity() {
```

Comment:
So I noticed a test above `test_enum_class_attr` with the following line:

```rust
Py::new(py, MyEnum::Variant).unwrap();
```

This is potentially a problem; that is a _new_ instance of `MyEnum` type and so it won't compare successfully with identity.

Now, the question is, is that a good thing? Pro could be that it gives users who need it an escape hatch. However, it is probably also a footgun.

Unfortunately, changing how `Py::new` works would require a refactoring of our initialization machinery. I'm sure there's plenty of scope for improvement in that area (e.g. #2384), however it would likely be a much bigger patch than this PR...

Reply:
I agree that this possibility is a footgun which we should prevent. Maybe we should put this a bit on hold until the scope for changes to `Py::new` is defined?

Reply:
By "this" do you mean solving this particular problem or the whole PR?

Reply:
I meant the whole pull request but I don't know what the best approach would be. It seems that comparing enums by identity is a desired feature but there are also two concerns https://github.com/PyO3/pyo3/pull/3061#discussion_r1150124372 and https://github.com/PyO3/pyo3/pull/3061#discussion_r1150128264 that would both require quite a refactor. Also happy to help in that regard but I don't know a lot about the existing problems and where to start. And I can imagine you don't want to complicate that work further by including some changes now?

Reply:
So we've just had #3287 merged which might be enough of a tweak to make this possible. Given that `Py::new` relies on `Into<PyClassInitializer<T>>` we might be able to adjust that trait implementation for enums to somehow use `PyClassInitializerImpl::Existing` (by fetching the value from a `GILOnceCell` which is itself initialized by a `PyClassInitializerImpl::New` invocation to `Py::new`.